docs/dev/rfcs/query/0007-semi-structured-data-types.md

```diff
@@ -0,0 +1,106 @@
+# Semi-structured data types design
+
+## Summary
+
+Semi-structured data types are used to represent schemaless data formats, such as JSON, XML, and so on.
+In order to be compatible with [Snowflake's SQL syntax](https://docs.snowflake.com/en/sql-reference/data-types-semistructured.html), we support the following three semi-structured data types:
+
+- `Variant`: A tagged universal type, which can store values of any other type, including `Object` and `Array`.
+- `Object`: Used to represent collections of key-value pairs, where the key is a non-empty string, and the value is a value of `Variant` type.
+- `Array`: Used to represent dense or sparse arrays of arbitrary size, where the index is a non-negative integer (up to 2^31-1), and values are `Variant` types.
+
+Since `Object` and `Array` can be regarded as a type of `Variant`, the following introduction mainly takes `Variant` as an example.
+
```

Comment:
Can you give an example of how users will use this feature? Like presenting a real SQL?


Take https://github.com/datafuselabs/opendal/blob/main/docs/rfcs/0000-example.md#guide-level-explanation for a look.

Reply:
I have added some examples, PTAL @Xuanwo